### ISSUE TITLE
Error  in "Step 2: Run the core Benisse "

### ISSUE THREAD

**Issue by Even-bio** [wooyongc/Benisse#12]
Hello, thank you for your contributions in the field of BCR sequencing and B cell transcriptional profiling. I would like to ask for your advice on a problem I'm currently facing. I am trying to run the second step "Step 2: Run the core Benisse model" using the example data provided in the installation package on GitHub (10x_NSCLC_exp.csv, 10x_NSCLC_contigs.csv, encoded_10x_NSCLC.csv). However, I am encountering the following error message when running the code
`Rscript Benisse.R \
example/10x_NSCLC_exp.csv \
example/10x_NSCLC_contigs.csv \
example/encoded_10x_NSCLC.csv \
example \
1610 1 100 1 1 10 1e-10
`
Upon running the code, I encountered the following error message:
`Loading required package: igraph
Attaching package: ‘igraph’
The following objects are masked from ‘package:stats’:
    decompose, spectrum
The following object is masked from ‘package:base’:
    union
[1] "Begin initiation."
[1] "Initiation success."
[1] "Iteration: 1"
Error in optim(as.vector(A[par_which]), fn = fn, gr = gr, method = "L-BFGS-B",  :
  L-BFGS-B needs finite values of 'fn'
Calls: Benisse -> update_A -> optim
Execution halted
`
Thanks!

**Comment by wooyongc** [comment]
Hello,

Thank you for your interest in our software. 
Unfortunately, the error was not replicable in my local device and is most likely a package issue.

I recommend that you create a separate R environment using renv or Conda.

**Comment by Even-bio** [comment]
Hello， I wanted to express my gratitude for your assistance .

Following your advice, I took the time to create a new R virtual environment. However, I still encountered the same error.It wasn't until I realized that when running the following code in "Step 1: Generating BCR `embedding":
```
python3 AchillesEncoder.py \
--input_data example/10x_NSCLC.csv \
--output_data example/encoded_10x_NSCLC.csv \
--cuda True
```

The **encoded_10x_NSCLC.csv** generated **was different from the one you provided in the "example" directory**. It's precisely the ****encoded_10x_NSCLC.csv**  that I generated**, which leads to the following error when proceeding to "Step 2: Run the core Benisse model." However, when using the **encoded_10x_NSCLC.csv** provided in your "example" directory, this error does not occur. 
```
Loading required package: igraph

Attaching package: ‘igraph’

The following objects are masked from ‘package:stats’:

decompose, spectrum

The following object is masked from ‘package:base’:

union

[1] "Begin initiation."
[1] "Initiation success."
[1] "Iteration: 1"
Error in optim(as.vector(A[par_which]), fn = fn, gr = gr, method = "L-BFGS-B",  : 
  L-BFGS-B needs finite values of 'fn'
Calls: Benisse -> update_A -> optim
Execution halted
```
In addition to the issue mentioned above, I would like to provide you with some information regarding the versions of packages and software on my computer:

- Python: 3.9.9
- PyTorch: 1.11.0
- Pandas: 1.3.5
- Scikit-Learn: 1.3.1 (I am aware that this version is deprecated)
- NumPy: 1.22.0
- R: 4.2.1
- ggplot2: 3.4.3
- data.table: 1.14.8
- igraph: 1.5.0

I have attached the  **encoded_10x_NSCLC.csv** generated by me after completing "Step 1: Generating BCR embedding" below for reference.Thank you once again for your assistance. Looking forward to your response.
[encoded_10x_NSCLC_mine.csv](https://github.com/wooyongc/Benisse/files/12847630/encoded_10x_NSCLC_mine.csv)

**Comment by wooyongc** [comment]
The encoding is expected to be a little different every time. 

I suggest that you install the software versions stated in our [dependency section.](https://github.com/wooyongc/Benisse#dependencies)

**Comment by Even-bio** [comment]
Hi!
I have installed the software versions stated in your dependency section.
**Python:3.7.8,  pytorch:1.10.0,  pandas:1.3.4, sklearn:1.0,  numpy : 1.21.3**
My encoded_10x_NSCLC.csv file is also having the same problem. 

`encoded_10x_NSCLC_mine$index[1:10]：
 [1] "<map object at 0x7f8ca19c9150>"
 [2] "<map object at 0x7f8ca19c9150>"
 [3] "<map object at 0x7f8ca19c9150>"
 [4] "<map object at 0x7f8ca19c9150>"
 [5] "<map object at 0x7f8ca19c9150>"
 [6] "<map object at 0x7f8ca19c9150>"
 [7] "<map object at 0x7f8ca19c9150>"
 [8] "<map object at 0x7f8ca19c9150>"
 [9] "<map object at 0x7f8ca19c9150>"
[10] "<map object at 0x7f8ca19c9150>"
`
Meanwhile，my encoded_10x_NSCLC.csv file is disorganized.
`
head(encoded_10x_NSCLC_mine)[1:10]
    X         X0         X1        X2         X3         X4       X5
1 12 -0.6136553 0.84690570 -1.948102 -0.5407202 -1.0677441 1.200367
2 17 -1.3373775 0.41448945 -2.234170 -0.4376893 -1.4455698 1.300224
3 27 -2.2516468 0.29882348 -2.453139 -0.0233840 -2.0565152 1.799484
4 62 -0.2426069 0.56131345 -1.647836 -1.1177971 -1.3327315 2.432032
5 32 -1.6620061 0.01906532 -1.629389 -0.9409735 -1.4806178 1.608033
6 22 -1.7615902 0.25832355 -2.774785 -1.4485345 -0.1137015 2.728931
         X6        X7         X8
1 -2.429961 -1.561266 -0.7867911
2 -3.511687 -1.722234 -0.7023264
3 -3.864463 -1.767459 -0.8134422
4 -0.808368 -2.649051 -0.3214853
5 -1.961402 -2.653506 -0.3482661
6 -1.537952 -1.392870 -0.7845711
`
Please debug！Thanks!
[encoded_10x_NSCLC_mine.csv](https://github.com/wooyongc/Benisse/files/12895332/encoded_10x_NSCLC_mine.csv)

**Comment by ncborcherding** [comment]
I encountered the same problem - I was able to track it down to the update.R - both fn() and gr() functions call in update_A() will return only NA if there are any NA values in the matrix. I was able to get things working with the following updates in the update.R

```
#Old Function
 #fn=function(par0,data){
  #  par=data[[10]]
  #  par[data[[9]]]=par0
  #  LA=as.vector(diag(rowSums(par))-par)
  #  0.5*data[[1]]*norm(data[[5]]-data[[8]]-4*data[[2]]*
  #                       (data[[6]]*data[[3]]+LA)-data[[4]]/data[[1]],type='F')^2+
  #    sum(par*as.vector(data[[7]]))
  #}
  
  fn=function(par0,data){
    par = data[[10]]
    par[data[[9]]] = par0
    LA = as.vector(diag(rowSums(par)) - par)
    
    # Handle NA values in data
    data[[5]][is.na(data[[5]])] <- 0
    data[[8]][is.na(data[[8]])] <- 0
    data[[6]][is.na(data[[6]])] <- 0
    data[[3]][is.na(data[[3]])] <- 0
    data[[4]][is.na(data[[4]])] <- 0
    data[[7]][is.na(data[[7]])] <- 0
    
    component1 = data[[5]] - data[[8]] - 4 * data[[2]] * (data[[6]] * data[[3]] + LA) - data[[4]] / data[[1]]
    norm_component1 = norm(component1, type = 'F')^2
    
    # Check for NA values in par
    par[is.na(par)] <- 0
    
    component2 = sum(par * as.vector(data[[7]]))
    
    result = 0.5 * data[[1]] * norm_component1 + component2
  }
  #Old Function
  #gr=function(par0,data){
  #  par=data[[10]]
  #  par[data[[9]]]=par0
  #  U=as.vector(diag(rowSums(par))-par)-
  #    (data[[5]]-data[[8]]-data[[4]]/data[[1]])/(4*data[[2]])+
  #    data[[6]]*data[[3]]
  #  gradients=2*data[[7]]+16*data[[1]]*data[[2]]^2*(-U-t(U)+2*diag(U))
  #  gradients[data[[9]]]
  #}
  
  gr = function(par0, data) {
    par = data[[10]]
    par[data[[9]]] = par0
    U = as.vector(diag(rowSums(par)) - par) -
      (data[[5]] - data[[8]] - data[[4]] / data[[1]]) / (4 * data[[2]]) +
      data[[6]] * data[[3]]
    
    # Handle NA values in data
    data[[7]][is.na(data[[7]])] <- 0
    
    gradients = 2 * data[[7]] + 16 * data[[1]] * data[[2]]^2 * (-U - t(U) + 2 * diag(U))
    gradients_subset = gradients[data[[9]]]
  }
```
